src/main/java/com/emc/pravega/perf/PravegaPerfTest.java

```diff
@@ -400,76 +602,70 @@ public void run() {
                 return null;
 
           };
-        }
-    }
+      }
+}
 
-    /**
-     * A Sensor reader class that reads the temperative data
-     */
-    private static class ReaderWorker implements Runnable {
-       @Setter
-        public static AtomicInteger totalEvents;
-        private EventStreamReader<String> reader;
-        String readerId;
+class ReaderWorker implements Callable<Void> {
+       public static AtomicInteger totalEvents;
+       private EventStreamReader<String> reader;
+       private final int secondsToRun;
+       private final long StartTime;
+       private final int timeout;
+       private final PerfStats stats;
+       String readerId;
 
 
-        public ReaderWorker(int readerId) {
+       ReaderWorker (int readerId, int secondsToRun, long start,
+                    ClientFactory factory, PerfStats stats,
+                    String readergrp, int timeout) {
             this.readerId = Integer.toString(readerId);
-            ClientFactory clientFactory = ClientFactory.withScope("Scope",
-                    URI.create(controllerUri));
-
-            reader = clientFactory.createReader(
-                    this.readerId, streamName, new JavaSerializer<String>(), ReaderConfig.builder().build());
+            this.secondsToRun = secondsToRun; 
+            this.StartTime=start;
+            this.stats = stats; 
+            this.timeout = timeout;
+
+            reader = factory.createReader(
+                    this.readerId, readergrp, new JavaSerializer<String>(), ReaderConfig.builder().build());
+       }
+
+       void cleanupEvents(PerfStats drainStats) throws Exception {
+            EventRead<String> result;
+            //System.out.format("******** Draining events from %s/%s%n", scopeName, streamName);
+            do {
+                long startTime = System.currentTimeMillis();
+                result = reader.readNextEvent(timeout);
+                if(result.getEvent()!=null) {
+                    drainStats.runAndRecordTime(() -> {
```

Comment:
why do we need a lambda function that just returns null?

Reply:
because runAnd Recordtime check if its null ; then calls record method to store the time stamps